kv/kvserver: TestReplicateQueueRebalanceMultiStore failed

kv/kvserver.TestReplicateQueueRebalanceMultiStore [failed](https://tanzanite.cluster.engflow.com/invocations/default/7bfb1ad6-f98c-4c40-bfa8-869678b7e8cb?testReportRun=22&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [75f19bec66a542d58680e9ccb47d57598f48fe20](https://github.com/cockroachdb/cockroach/commits/75f19bec66a542d58680e9ccb47d57598f48fe20):


```
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [31 34 35 32 30 33 31 29]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 21 7 9 3 7 11 7]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [31 34 35 32 30 33 31 29]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 21 7 9 3 7 11 7]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [31 34 35 32 30 33 31 29]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 21 7 9 3 7 11 7]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [31 34 35 32 30 33 31 29]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 21 7 9 3 7 11 7]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [31 34 35 32 30 33 31 29]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 21 7 9 3 7 11 7]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [31 34 35 32 30 33 31 29]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 21 7 9 3 7 11 7]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [31 34 35 32 30 33 31 29]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 21 7 9 3 7 11 7]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [31 34 35 32 30 33 31 29]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 21 7 9 3 7 11 7]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [31 34 35 32 30 33 31 29]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 21 7 9 3 7 11 7]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [31 34 35 32 30 33 31 29]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 21 7 9 3 7 11 7]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [31 34 35 31 30 33 32 29]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 21 7 9 3 7 11 7]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [31 34 35 31 30 33 33 29]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 21 7 9 3 7 11 7]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [31 34 35 30 30 33 33 29]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [17 21 7 9 3 7 11 7]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [31 34 35 29 29 33 33 30]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [17 21 7 9 3 7 11 8]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [32 34 35 29 28 33 33 31]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [17 21 7 9 3 7 11 8]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [34 33 34 29 28 33 33 31]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [17 20 7 8 5 7 10 9]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [34 33 33 29 31 33 32 31]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 20 6 8 6 7 9 9]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [34 33 32 29 32 33 32 31]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 19 6 8 8 7 8 9]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [34 32 31 29 33 33 32 31]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 19 6 8 8 7 8 9]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [34 31 31 29 35 33 31 31]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 19 6 8 9 6 8 9]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [34 31 31 29 35 33 31 31]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 19 6 8 9 6 8 9]
    replicate_queue_test.go:329: current replica state (want at least 28 replicas on all stores): [34 30 31 29 35 30 31 31]
    replicate_queue_test.go:340: current lease state (want at least 7 leases on all stores): [18 18 7 8 9 7 8 8]
    replicate_queue_test.go:355: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:355
        	Error:      	Received unexpected error:
        	            	pq: inbox communication error: rpc error: code = Canceled desc = context canceled
        	Test:       	TestReplicateQueueRebalanceMultiStore/multi-store
    --- FAIL: TestReplicateQueueRebalanceMultiStore/multi-store (450.47s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=22</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicateQueueRebalanceMultiStore.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-35361

kv/kvserver.TestReplicateQueueRebalanceMultiStore [failed](https://tanzanite.cluster.engflow.com/invocations/default/5f3706aa-b386-4ef7-9c64-d08ed1548e7e?testReportRun=27&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [1898adc059e9b3026fd5a7ed88dc41cbb006abfc](https://github.com/cockroachdb/cockroach/commits/1898adc059e9b3026fd5a7ed88dc41cbb006abfc):


```
=== RUN   TestReplicateQueueRebalanceMultiStore
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestReplicateQueueRebalanceMultiStore2168183203
    test_log_scope.go:81: use -show-logs to present logs inline
    replicate_queue_test.go:367: -- test log scope end --
test logs left over in: outputs.zip/logTestReplicateQueueRebalanceMultiStore2168183203
--- FAIL: TestReplicateQueueRebalanceMultiStore (649.34s)
=== RUN   TestReplicateQueueRebalanceMultiStore/multi-store
    replicate_queue_test.go:273: split off r68:/{Table/100-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=9, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:273: split off r69:/{Table/101-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=10, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:273: split off r70:/{Table/102-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=11, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:273: split off r71:/{Table/103-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=12, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:273: split off r72:/{Table/104-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=13, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:273: split off r73:/{Table/105-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=14, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:273: split off r74:/{Table/106-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=15, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:273: split off r75:/{Table/107-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=16, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:273: split off r76:/{Table/108-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=17, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:273: split off r77:/{Table/109-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=18, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:273: split off r78:/{Table/110-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=19, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:273: split off r79:/{Table/111-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=20, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:273: split off r80:/{Table/112-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=21, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:273: split off r81:/{Table/113-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=22, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:273: split off r82:/{Table/114-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=23, sticky=9223372036.854775807,2147483647]
    replicate_queue_test.go:259: condition failed to evaluate within 45s: from testserver.go:2072: /Table/115: split unexpected error: replica unavailable: (n3,s6):3 unable to serve request to r82:/{Table/114-Max} [(n4,s7):1, (n2,s4):2, (n3,s6):3, next=4, gen=23, sticky=9223372036.854775807,2147483647]: closed timestamp: 1705471598.729411517,0 (2024-01-17 06:06:38); raft status: {"id":"3","term":64,"vote":"3","commit":24,"lead":"3","raftState":"StateLeader","applied":24,"progress":{"1":{"match":0,"next":40,"state":"StateProbe"},"2":{"match":0,"next":40,"state":"StateProbe"},"3":{"match":40,"next":41,"state":"StateReplicate"}},"leadtransferee":"0"}: have been waiting 61.00s for slow proposal RequestLease [/Table/114]
    --- FAIL: TestReplicateQueueRebalanceMultiStore/multi-store (604.46s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=27</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicateQueueRebalanceMultiStore.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/replicate_queue_test.go
+++ b/pkg/kv/kvserver/replicate_queue_test.go
@@ -221,6 +221,11 @@ func TestReplicateQueueRebalanceMultiStore(t *testing.T) {
 	for _, testCase := range testCases {
 
 		t.Run(testCase.name, func(t *testing.T) {
+			if testCase.storesPerNode > 1 {
+				// 8 stores with active rebalancing can lead to failed heartbeats due
+				// to overload. Skip under stress when running the multi-store variant.
+				skip.UnderStress(t)
+			}
 			// Set up a test cluster with multiple stores per node if needed.
 			args := base.TestClusterArgs{
 				ReplicationMode:   base.ReplicationAuto,

